Asyncronous tasks

AlejandroRG31 opened the pull request:
- Installed background_task and created a file tasks.py to execute importation of data, variation and gramcats in the background.

- Moved part of exportdata.py, exportvariation.py, importdata.py, importvariation.py and importgramcats.py to importers.py and exporters.py file to execute them after have been called from tasks.py

- Added the views of importers and exporters to allow use them from the a browser and update the database or download information in a more clear way.

- Added a new model ImportersInfo that keep track of the current state of the importers and save which errors have it or not after its execution.

- Added two views: one to see all the importations made and another one (which can be accessed from the first one), with more info about a specific importation and its errors.

TO DO:

- [x] Modify the current way of how the errors of importations are stored or at least how they are shown (Now is a string with all the errors which are separated calling errors.split( }, ))

- [x] Add a way to keep track of errors such as IntegrityError...

- [x] Solve the error that happens when trying to execute the tests (I need help with this)

- [ ] Change MAX_ATTEMPS in background_tasks to a smaller number like 2, instead of 25.

- [ ] ExportData.html now it not being divided in two columns. (But maybe it is better to be displayed in rows)

Attached some screenshots to see how it looks now:

![image](https://user-images.githubusercontent.com/64277128/102022323-596a9000-3d86-11eb-953a-2c67bb6ca3d9.png)

![image](https://user-images.githubusercontent.com/64277128/102022311-46f05680-3d86-11eb-94ea-7426eb5c6584.png)

![image](https://user-images.githubusercontent.com/64277128/102396955-8664b000-3fdd-11eb-939e-9f398af95984.png)

AlejandroRG31 committed:
Added django asyncronic tasks which uses file tasks.py to execute the exporters

AlejandroRG31 committed:
Added the methods in tasks.py, now are called with .now() and created space in views for ExportData view

AlejandroRG31 committed:
Added templates to download export files

AlejandroRG31 committed:
Fixed some problems when writting in variation file, now works perfectly using .now()

AlejandroRG31 committed:
Added installed app

AlejandroRG31 committed:
Added django-background-tasks in github test

AlejandroRG31 committed:
Trying differents requirements version for testing

AlejandroRG31 committed:
Trying diferent version of requirements for testing

AlejandroRG31 committed:
Tried diferents version requirements for testing

AlejandroRG31 committed:
Trying diferent python version for testing

AlejandroRG31 committed:
Changed django version for testing

AlejandroRG31 committed:
Fixed error in exportvariation and added export gramcats in tasks

AlejandroRG31 committed:
Updated requirements to previous django version

AlejandroRG31 committed:
Updated github test to previous python version

AlejandroRG31 committed:
Merge remote-tracking branch 'origin/feature/asyncronic-tasks' into feature/asyncronic-tasks

AlejandroRG31 committed:
Merge branch 'master' into feature/asyncronic-tasks

AlejandroRG31 committed:
Moved logic of export and import files to other files, added view to import DiatopicVariation.

Moved logic of importing or exporting in the database to importers.py and exporters.py and deleted from tasks.py.

Added a view importdata to use all the importers in only one view.

Added importation view to import variation using the browser, works fine also in an asyncronic way.

AlejandroRG31 committed:
Added model that allows keep track of importations and added data importer and views to see both

Model importsinfo save the info of how is the current state of an importation and its errors

Added the last importer importdata to the tasks file with its view added to imports view

Added a view that allows see all importations and select one to go to a specific one

Added another view to see all the info and errors of an importation

AlejandroRG31 committed:
Fixed two bugs

Status now is shonw in the view page

Now Integrity errors are stored in importationinfo but background-tasks keeps executing it again and again

AlejandroRG31 committed:
Fixed the number of errors that wasn't giving the correct number

slamora committed:
Not disable contenttype migration.

background task libary has a dependency.

slamora committed:
Keep MIGRATION_MODULES emty.

AlejandroRG31 committed:
Solved some errors pointed out by the tests

Added 2 new methods to save the data in InfoImporters model in a more clear way

Now variation variable is passed as name instead of pk

Solved an errors which was causing words to be saved even when they had errors

Removed COMPLETED_WITH_ERRORS in ImportsInfo model which now is not necessary

AlejandroRG31 committed:
Added xlrd==1.2.0 boto3 because of a xlsx error

AlejandroRG31 committed:
Now error are displayed in a more clear way

Also added the error IntegrityError in importers.py variation

Schedule time changed from 30 to 0

AlejandroRG31 committed:
Merge branch 'feature/asyncronic-tasks' of https://github.com/ribaguifi/linguatec-lexicon into feature/asyncronic-tasks

AlejandroRG31 committed:
Now also catch IntegrityError in importers.py load_gramcats

slamora commented on linguatec_lexicon/models.py:
```diff
@@ -265,3 +267,50 @@ def model_word_id(self):
+        return errors_list
+
+
+class ImportsInfo(AbstractImportsInfo):
```
You can create directly `ImportsInfo` model without using an intermediary abstract class. Besides that my proposal is to rename this class to `ImportLog`

AlejandroRG31 replied:
Changed, definitely it is better this way, thank you.

slamora commented on linguatec_lexicon/templates/linguatec_lexicon/exportdata.html:
```diff
@@ -0,0 +1,52 @@
+
+{% block content %}
+
+<div style="width: 45%; float:left">
```
Instead of withd and float, I think that it's better to use **bootstrap layout** (rows + cols). I'm not sure which version is included with django-rest-framework.

AlejandroRG31 replied:
Trying to use a contatiner with one row and two columns but for whatever reason it is not working, but maybe could be better to be displayed in rows

slamora commented on linguatec_lexicon/templates/linguatec_lexicon/importdata.html:
```diff
@@ -0,0 +1,76 @@
+<div class="row">
+    <div class="col-12">
+        <div>
+            Selecciona el fichero XLSX para importar las categorias gramaticales.
```
Typo: `categorías` (tilde missing )

AlejandroRG31 replied:
Fixed

slamora commented on linguatec_lexicon/templates/linguatec_lexicon/importdata.html:
```diff
@@ -0,0 +1,76 @@
+<div class="row">
+    <div class="col-12">
+        <div>
+            Selecciona el fichero XLSX para importar las variaciones diatopicas.
```
`diatópicas` (tilde missing)

AlejandroRG31 replied:
Fixed

slamora commented on linguatec_lexicon/views.py:
```diff
@@ -75,11 +79,182 @@ def validate(self, xlsx_file):
+
+        type_import = str(request.POST.get('type_import'))
+
+        if type_import == 'data' or type_import == 'variation':
```
Please define constants instead of hardcode `data`, `variation`... 
Maybe you can define it on the model ImportsInfo (ImportLog) and then access to it outside using something like: `ImportLog.TYPE_DATA`

You can check [django-importer](https://github.com/ribaguifi/django-importer/blob/master/djimporter/models.py#L6-L16) to get some inspiration.

AlejandroRG31 replied:
Corrected, now is used the constants already defined instead of hardcoding them each time, thank you.

slamora commented on linguatec_lexicon/views.py:
```diff
@@ -75,11 +79,182 @@ def validate(self, xlsx_file):
+
+            if type_import == 'data':
+                lexicon_code = str(request.POST.get('lexicon_code'))
+                lexicon_id = Lexicon.objects.get(src_language=lexicon_code[:2], dst_language=lexicon_code[3:]).pk
```
I suggest to create a custom method on LexiconManager to retrieve a Lexicon by its code. I mean something like that:
```python
# models.py
class LexiconManager(models.Manager):
    def get_by_code(self, code):
        src_lang = code[:2]
        dst_lang = code[3:]
        return self.get(src_language=src_lang, dst_language=dst_lang)

class Lexicon(models.Model):
    ....
    objects = LexiconManager()
```

You can get more information checking the docs of [custom model managers](https://docs.djangoproject.com/en/3.1/topics/db/managers/#custom-managers).

AlejandroRG31 replied:
Added and changed all the code to use .get_by_code instead of appliyng and rewritting this function in all the code, thank you for the info and aportation

dunetna commented on linguatec_lexicon/exporters.py:
```diff
@@ -0,0 +1,118 @@
+
+        writer = csv.DictWriter(outfile, fieldnames=fieldnames, delimiter=';')
+
+        # word_list = []
```
Remove commented unused line

AlejandroRG31 replied:
Removed, thank you for finding it

AlejandroRG31 committed:
Modified models, fixed some gramatical errors and modified exportdata.html

Modified Word models adding a lexicon manager to get a lexicon using .get_by_code method.

Changed name ImportsInfo to ImportLog in all files and converted to a not abstract model.

Now ImportLog.TYPE_DATA is used instead of hardcoding the strings.

Fixed some typos.

Changed exportdata.html to use bootstrap to be divided in columns.

dunetna commented on linguatec_lexicon/importers.py:
```diff
@@ -0,0 +1,457 @@
+from linguatec_lexicon.models import (Entry, GramaticalCategory, Word,
+                                      Lexicon, DiatopicVariation, Example,
+                                      VerbalConjugation, ImportsInfo)
+
```
Extra blank line. You can use "Sort imports" from VS to organize imports (right-click on the code)

AlejandroRG31 replied:
Organized and removed extra line, thank you for the tip

dunetna commented on linguatec_lexicon/importers.py:
```diff
@@ -0,0 +1,457 @@
+            # includes empty string. We don't want that pandas replaces empty
+            # cells with 'nan'
+            partial = xlsx.parse(sheet, header=None, usecols='A:F')
+            # names=['colA', 'colB', 'colC', 'colD', 'colE', 'colF'])
```
Remove unused code

AlejandroRG31 replied:
This was there before moving the code, I'm not sure if it is an unused line of code or just a comment

dunetna commented on linguatec_lexicon/importers.py:
```diff
@@ -0,0 +1,457 @@
+        if errors is not None:
+            imports_info.errors = errors
+        imports_info.save()
+
```
We should have 1 blank line between functions and 2 between classes (review all this file, some times there are two lines, sometimes one)

AlejandroRG31 replied:
Flake8 is not tracking those lines, i don't know if this could be because those methods are not class methods, so maybe it is correct or flake8 is not tracking them correctly

dunetna commented on linguatec_lexicon/templates/linguatec_lexicon/importdata.html:
```diff
@@ -0,0 +1,76 @@
+            <form action="" enctype="multipart/form-data" method="post">
+                {% csrf_token %}
+                {{ form.as_p }}
+                <label for="lexicon_code">Codigo lexicon</label><input type="text" name="lexicon_code" /></br>
```
Código (tilde missing)
Leave a blank space between Código and input.

AlejandroRG31 replied:
Fixed, thank you for pointing out it

dunetna commented on linguatec_lexicon/templates/linguatec_lexicon/importdata.html:
```diff
@@ -0,0 +1,76 @@
+            <form action="" enctype="multipart/form-data" method="post">
+                {% csrf_token %}
+                {{ form.as_p }}
+                <label for="lexicon_code">Codigo lexicon</label><input type="text" name="lexicon_code" /></br>
```
Código (tilde missing)
Leave a blank space between Código and input.

AlejandroRG31 replied:
Fixed

dunetna reviewed (commented):
* django-background-tasks should be added to requirements.txt
* importdata interface should be improved beautifyng forms (using bootstrap)

Good job!

AlejandroRG31 committed:
Added display lists, changed errors and fixed minor errors

Added display lists to select lexicon and variation in importdata.html

Changed a display error in ImportLog to show 0 errors when there are no errors

And fixed minor errors and cleaned the code

AlejandroRG31 committed:
Changed exportdata.html to use display lists instead of textfields

AlejandroRG31 committed:
Fixed error in validators

Added in validators view a display list to choose a lexicon and variation to do the validation

dunetna commented on linguatec_lexicon/templates/linguatec_lexicon/datavalidator.html:
```diff
@@ -18,6 +18,33 @@ <h1>{{ title }}</h1>
+                <div class="input-group rg-search">
+                    <label for="lexicon_code">Código lexicon</label>
+                    <select id="selected_lex" name="lexicon_code">
+                        {% for lexicon in lexicon_list %}
+                            <optgroup label="{{ lexicon.src_language }}"> 
+                            {% for lex in lexicon_list %}
+                                {% if lex.src_language == lexicon.src_language %}
+                                    <option value="{{ lex.code }}">{{ lex.code }}</option> 
+                                {% endif %}
+                            {% endfor %}
+                        {% empty %}
+                        <option selected value=''>Empty Value</option>
+                        {% endfor %}
```
These two nested loops are traversing the same list. Is that correct?
If it is correct, may we remove the nested loop, writing directly <option value="{{ lexicon.code }}">{{ lexicon.code }}</option>

AlejandroRG31 replied:
Yeah my bad, I wrote it a bit fast, the idea is to use the src_languages (appliyng a distinct) as labels to order the lexicon in the display list. I just have rewritten and fixed it, thank you for pointing it out.

AlejandroRG31 committed:
Refactored loop of src_languages to work correctly

AlejandroRG31 committed:
Eliminated ImportErrorsViewSet and some duplicated code

Eliminated ImportErrorsViewSet because it was returning a json page with all the errors of the importation and it is not necesary to implement this characteristic because it is not going to be used

AlejandroRG31 committed:
Fixed flake8 errors in commands
